File: cpp/sliding_window/best_time_to_buy_and_sell_stock_leetcode_121.cpp
#include <bits/stdc++.h>
using namespace std;

// Two Pointer Method O(N)
class Solution {
  
  public:
    int maxProfit(vector<int>& prices){
      int l = 0, r = 1;
      int maxP = 0;
      while(r < prices.size()){
        if(prices[l] < prices[r]){
          maxP = max(maxP, prices[r] - prices[l]);
        } else {
          l = r;
        }
        r++;
      }
    }
};

/* Using Dynamic Programming 

class Solution{
  public: 
    int maxProfit(vector<int>& prices){
      int maxP = 0;
      int minB = prices[0];
      for(int sel : prices){
        maxP = max(maxP, sel-minB);
        minB = min(sel, minB);
      }
      return maxP;
    }
};

*/ 

File: python/sliding_window/Best Time to Buy and Sell Stock_leetcode_121.py
''' My method Not very efficient, but works
class Solution(object):
    def maxProfit(self, prices):
        maxProfit = 0
        for price in prices:
            try:
                maxx = max(prices[prices.index(price)+1:])
                maxProfit = max(maxProfit, maxx - price)
            except:
                pass
        return maxProfit
'''

class Solution(object):
    def maxProfit(self, prices):
        l, r = 0, 1
        maxp = 0
        while l < r:
            if prices[l] < prices[r]:
                maxp = max(maxp, prices[r]-prices[l])
            else:
                l = r
            r += 1

if __name__ == "__main__":
    arr = [7,1,5,3,6,4]
    sol = Solution()
    ans = sol.maxProfit(arr)
    print(ans)
